### PR TITLE
RavenDB-8401 Added Esprima Visitor and support for LoadDocument in Ja…

### PR DESCRIPTION
**Pull request by talweiss1982**
…vaScriptIndex

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+                    VisitWithStatement(statement.As<WithStatement>());
+                    break;
+                case Nodes.Program:
+                    VisitProgram(statement.As<Esprima.Ast.Program>());
+                    break;
+                case Nodes.CatchClause:
+                    VisitCatchClause(statement.As<CatchClause>());
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs:87]: Add the type here, so we'll know what we missed

File: src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,767 @@
...
+                    VisitUpdateExpression(expression.As<UpdateExpression>());
+                    break;
+                case Nodes.UnaryExpression:
+                    VisitUnaryExpression(expression.As<UnaryExpression>());
+                    break;
+                case Nodes.ArrowFunctionExpression:
+                    VisitArrowFunctionExpression(expression.As<ArrowFunctionExpression>());
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs:295]: Add the type here so we'll know what we missed

File: src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,767 @@
...
+                case Nodes.ExportAllDeclaration:
+                    VisitExportAllDeclaration(node.As<ExportAllDeclaration>());
+                    break;
+                case Nodes.ExportDefaultDeclaration:
+                    VisitExportDefaultDeclaration(node.As<ExportDefaultDeclaration>());
+                    break;
+                case Nodes.ClassExpression:
+                    VisitClassExpression(node.As<ClassExpression>());
+                    break;
+            }
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs:582]: Need a default here

File: src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,767 @@
...
+        public virtual void VisitImportSpecifier(ImportSpecifier importSpecifier)
+        {
+        }
+
+        public virtual void VisitMethodDefinition(MethodDefinition methodDefinitionas)
+        {
+        }
+
+        public virtual void VisitForOfStatement(ForOfStatement forOfStatement)
+        {
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs:626]: Missing implementation here.

Comment (talweiss1982) [reply src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs:626]: Jint doesn't seem to have support for this, it is part of the completion of the solution that i have added it, but i'll add an implementation to support it in the visitor.

**Pull request merged**